webapp/src/ts/services/transitions/create-user-for-contacts.transition.ts

```diff
@@ -0,0 +1,173 @@
+  private replaceForms;
+
+  /**
+   * @param {Object} settings - the CHT instance settings
+   * @return {Boolean} - whether the required config is present
+   */
+  init(settings) {
+    this.replaceForms = this.getReplaceForms(settings);
+    if (!this.replaceForms || !Array.isArray(this.replaceForms) || !this.replaceForms.length) {
+      console.warn(
+        `Configuration error. Config must define have a '${this.name}.${this.REPLACE_PROPERTY}' array defined.`
+      );
+
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @param {Array<Doc>} docs - docs to be saved
+   * @return {Boolean} - whether any of the docs from the batch should be processed
+   */
+  filter(docs: Doc[]) {
+    return !!docs.filter(doc => doc.type === 'data_record').length;
+  }
+
+  /**
+   * @param {Array<Doc>} docs - docs to run the transition over
+   * @returns {Promise<Array<Doc>>} - updated docs (may include additional docs)
+   */
+  async run(docs: Doc[]) {
+    const originalContact = await this.createUserForContactsService.getUserContact();
+    if (!originalContact) {
+      return docs;
+    }
+
+    if (this.createUserForContactsService.isReplaced(originalContact)) {
+      this.reparentReports(docs, originalContact);
+    }
+    const userReplaceDoc = this.getUserReplaceDoc(docs);
+    if (userReplaceDoc) {
+      return this.replaceUser(docs, userReplaceDoc, originalContact);
+    }
+
+    return docs;
+  }
+
+  private getTransitionConfig(settings) {
+    return settings[this.name];
+  }
+
+  private getReplaceForms(settings) {
+    return this.getTransitionConfig(settings)?.[this.REPLACE_PROPERTY] || [];
+  }
+
+  private getUserReplaceDoc(docs: Doc[]) {
+    return docs.find(doc => this.replaceForms.includes(doc.form)) as UserReplaceDoc;
+  }
+
+  private async replaceUser(docs: Doc[], userReplaceDoc: UserReplaceDoc, originalContact: Doc) {
+    const { replacement_contact_id } = userReplaceDoc.fields;
+    if (!replacement_contact_id) {
+      throw new Error('The form for replacing a user must include a replacement_contact_id field ' +
+        'containing the id of the new contact.');
+    }
+
+    const originalContactId = userReplaceDoc.contact._id;
+
+    const isOriginalContact = () => originalContact._id === originalContactId;
+    const isReplacedContact = () => this.createUserForContactsService.isReplaced(originalContact) &&
+      this.createUserForContactsService.getReplacedBy(originalContact) === originalContactId;
+    if (!isOriginalContact() && !isReplacedContact()) {
+      throw new Error('Only the contact associated with the currently logged in user can be replaced.');
+    }
+    const newContact = await this.getNewContact(docs, replacement_contact_id);
+    if (!newContact) {
+      throw new Error(`The new contact could not be found [${replacement_contact_id}].`);
+    }
+    this.createUserForContactsService.setReplaced(originalContact, newContact);
+    const updatedDocs = [...docs, originalContact];
```

Comment:
I always try to avoid mutating parameters as a best practice.  After considering your comment about about refactoring the `run` function, I agree with you that it makes more sense to just mutate `docs` in here (particularly since that is pretty much the original purpose for passing `docs` to run in the first place....).  I have changed this code to just mutate `docs`.